modules/saslserv/authcookie.c

```diff
@@ -11,44 +11,45 @@
 static const sasl_core_functions_t *sasl_core_functions = NULL;
 
 static int
-mech_step(sasl_session_t *p, char *message, size_t len, char **out, size_t *out_len)
+mech_step(sasl_session_t *const restrict p, char *const restrict message, const size_t len,
+          char __attribute__((unused)) **const restrict out, size_t __attribute__((unused)) *const restrict out_len)
 {
-	char authz[256];
-	char authc[256];
-	char cookie[256];
-	myuser_t *mu;
-
-	/* Skip the authzid entirely */
+	if (! (message && len))
+		return ASASL_FAIL;
 
-	if(strlen(message) > 255)
+	char data[768];
+	if (len >= sizeof data)
 		return ASASL_FAIL;
-	len -= strlen(message) + 1;
-	if(len <= 0)
+
+	(void) memset(data, 0x00, sizeof data);
+	(void) memcpy(data, message, len);
+
+	const char *ptr = data;
+	const char *const end = data + len;
+
+	const char *const authzid = ptr;
+	if (! *ptr || (ptr += strlen(ptr) + 1) >= end)
 		return ASASL_FAIL;
-	strcpy(authz, message);
-	message += strlen(message) + 1;
 
-	/* Copy the authcid */
-	if(strlen(message) > 255)
+	const char *const authcid = ptr;
+	if (! *ptr || (ptr += strlen(ptr) + 1) >= end)
```

Comment:
Really not sure whether I like assignments hidden so deep inside an `if()`.

Reply:
The alternative is a lot more lines that do the same thing; a la

```
    if (! *ptr)
        return ASASL_FAIL;

    ptr += strlen(ptr) + 1;
    if (ptr >= end)
        return ASASL_FAIL;
```

That would get old, quickly, no?

Reply:
Eh, I think that'd be fine. Yes, I guess you can't fit as much in a single screen, but (as someone who isn't all that fluent in C) it's easier to decipher at a quick glance – specifically, it's easier to notice that `ptr` is being assigned to.

`||` chains are fine, assignments inside if() are okay I suppose, but _combining_ the two begins to look like golfing just for the sake of it... How about this then:

``` c
const char *const authzid = ptr;

if (! *ptr)
    return ASASL_FAIL;
if ((ptr += strlen(ptr) + 1) >= end)
    return ASASL_FAIL;

const char *const authcid = ptr;
```

(Maybe that first check should in fact be `if (! *authzid)`? Technically it's the same thing, but semantically it makes more sense.)

Reply:
How about

```C
const char *const authzid = ptr;

if (! *authzid)
    return ASASL_FAIL;
if ((ptr += strlen(authzid) + 1) >= end)
    return ASASL_FAIL;

const char *const authcid = ptr;

// ...
```

Reply:
LGTM

Reply:
https://github.com/atheme/atheme/pull/598/commits/8c64dce849a0ec6003e8f9b0874d4ef395a8bb3c